### PR TITLE
[rllib] Refactor rllib to have a common sample collection pathway

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Currently RLlib algorithms have disparate sample collection pathways. This makes supporting common functionality such as LSTMs, env vectorization, batch norm, and multi-agent hard to do in a generic way.

This PR adds a `CommonPolicyEvaluator` class which is responsible for routing observations to `Policy` and `TFPolicy` instances. In the multi-agent case, this many involve batching and routing observations to several local policies. It will also handle the vectorized env case.

- [x] A3C
- [x] PG
- [x] DQN
- [x] DDPG
- [x] punting on PPO, DDPG2 for now
- [x] validate learning performance did not regress
- [x] unit tests for CommonPolicyEvaluator
- [x] Rename policy loss to policy graph

## Related issue number

https://github.com/ray-project/ray/issues/2053

### CODE REVIEW

**Review by richardliaw — COMMENTED** [review]
partial read

**Comment by alok** [comment]
Since this would end up changing pretty much all of rllib, perhaps now is a good time to add rllib to the yapf linter. That way dataframe is the only code that doesn't get autoformatted.

**Comment by richardliaw** [comment]
I think linting PRs should be separate - and we can have one immediately after this. 

Coupling core refactoring and linting would just make things impossible to review...

**Comment by ericl** [comment]
Oh yeah, let's please hold off on linting for rllib for the next few months. We can do lint once things are stabilized.

File: python/ray/rllib/utils/process_rollout.py [hunk 1]
```diff
@@ -32,20 +32,19 @@ def process_rollout(rollout, reward_filter, gamma, lambda_=1.0, use_gae=True):
...
     else:
-        rewards_plus_v = np.stack(rollout["rewards"] +
-                                  [np.array(rollout.last_r)]).squeeze()
+        rewards_plus_v = np.concatenate(
+            [rollout["rewards"], np.array([last_r])])
         traj["advantages"] = discount(rewards_plus_v, gamma)[:-1]
 
     for i in range(traj["advantages"].shape[0]):
-        traj["advantages"][i] = reward_filter(traj["advantages"][i])
+        traj["advantages"][i] = traj["advantages"][i]
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/process_rollout.py]: Removed

File: python/ray/rllib/utils/tf_policy_loss.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_loss.py]: Done

File: python/ray/rllib/optimizers/sample_batch.py [hunk 3]
```diff
@@ -2,17 +2,22 @@
...
-    else:
-        return np.array(s)
+class SampleBatchBuilder(object):
+    """Util to build a SampleBatch incrementally."""
+
+    def __init__(self):
+        self.buffers = collections.defaultdict(list)
+
+    def add_values(self, **values):
+        for k, v in values.items():
```
Comment (alok) [inline python/ray/rllib/optimizers/sample_batch.py]: This would probably be easier to write as `self.buffers.update(values)`

Comment (ericl) [reply python/ray/rllib/optimizers/sample_batch.py]: That does not have the same semantics.

File: python/ray/rllib/dqn/dqn.py [hunk 4]
```diff
@@ -137,16 +120,31 @@ def default_resource_request(cls, config):
...
+            compress_observations=True,
+            registry=self.registry, env_config=self.config["env_config"],
+            model_config=self.config["model"], policy_config=self.config)
+        remote_cls = CommonPolicyEvaluator.as_remote(
             num_cpus=self.config["num_cpus_per_worker"],
-            num_gpus=self.config["num_gpus_per_worker"])(
-            DQNEvaluator)
+            num_gpus=self.config["num_gpus_per_worker"])
         self.remote_evaluators = [
             remote_cls.remote(
```
Comment (richardliaw) [inline python/ray/rllib/dqn/dqn.py]: it might even be cleaner if 
```

**Review by richardliaw — COMMENTED** [review]
Left some questions. Do we have a list or issue where all the refactoring is centered around?

General list of later todos:
 - PyTorch Policy Graph
 - Moving PPO onto the common evaluator
 - consider a better way of managing exploration
 - managing filters? 

Other nit:
 - Perhaps consider exposing something instead of LocalSyncReplay - something that puts the `for loop` of evaluation up front, and then think about the process of going from single thread to multi-process/multi-machine and making that process easy to do.

File: python/ray/rllib/dqn/dqn.py [hunk 5]
```diff
@@ -177,69 +190,34 @@ def _train(self):
...
-            explorations.append(s["exploration"])
-
-        opt_stats = self.optimizer.stats()
-
-        result = TrainingResult(
-            episode_reward_mean=mean_100ep_reward,
-            episode_len_mean=mean_100ep_length,
-            episodes_total=num_episodes,
-            timesteps_this_iter=self.global_timestep - start_timestep,
+        exp_vals = [self.exploration0.value(self.global_timestep)]
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/dqn/dqn.py]: Hmm, if we expose some "global stats" object then it could.

**Comment by ericl** [comment]
> Perhaps consider exposing something instead of LocalSyncReplay - something that puts the for loop of evaluation up front, and then think about the process of going from single thread to multi-process/multi-machine and making that process easy to do.

You can always copy paste the code and run it directly right? I don't think policy optimizers are required unless your actually putting your algorithm into rllib.

File: python/ray/rllib/a3c/a3c_evaluator.py [hunk 6]
```diff
@@ -1,119 +0,0 @@
...
-            registry, env_creator(config["env_config"]), config["model"])
-        self.env = env
-        policy_cls = get_policy_cls(config)
-        # TODO(rliaw): should change this to be just env.observation_space
-        self.policy = policy_cls(
-            registry, env.observation_space.shape, env.action_space, config)
-        self.config = config
-
-        # Technically not needed when not remote
-        self.obs_filter = get_filter(
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/a3c/a3c_evaluator.py]: They should be

File: python/ray/rllib/a3c/tfpolicy.py [hunk 7]
```diff
@@ -1,106 +0,0 @@
...
-            tf.summary.scalar("model/policy_loss", self.pi_loss / bs)
-            tf.summary.scalar("model/value_loss", self.vf_loss / bs)
-            tf.summary.scalar("model/entropy", self.entropy / bs)
-            tf.summary.scalar("model/grad_gnorm", tf.global_norm(self.grads))
-            tf.summary.scalar("model/var_gnorm", tf.global_norm(self.var_list))
-            self.summary_op = tf.summary.merge_all()
-
-        # TODO(rliaw): Can consider exposing these parameters
-        self.sess = tf.Session(graph=self.g, config=tf.ConfigProto(
-            intra_op_parallelism_threads=1, inter_op_parallelism_threads=2,
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/a3c/tfpolicy.py]: Fixed

File: python/ray/rllib/bc/policy.py [hunk 8]
```diff
@@ -64,7 +56,7 @@ def setup_gradients(self):
     def initialize(self):
         if self.summarize:
             bs = tf.to_float(tf.shape(self.x)[0])
-            tf.summary.scalar("model/policy_loss", self.pi_loss / bs)
+            tf.summary.scalar("model/policy_graph", self.pi_loss / bs)
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/bc/policy.py]: Oops fixed

File: python/ray/rllib/ppo/ppo_evaluator.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 from ray.rllib.utils.sampler import SyncSampler
 from ray.rllib.utils.filter import get_filter, MeanStdFilter
 from ray.rllib.utils.process_rollout import process_rollout
-from ray.rllib.ppo.loss import ProximalPolicyLoss
+from ray.rllib.ppo.loss import ProximalPolicyGraph
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/ppo/ppo_evaluator.py]: Yeah, we should do that.

File: python/ray/rllib/test/test_supported_spaces.py [hunk 10]
```diff
@@ -36,32 +36,6 @@
         Box(0.0, 1.0, (5,), dtype=np.float32)]),
 }
 
-# (alg, action_space, obs_space)
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/test/test_supported_spaces.py]: I'm just throwing unsupported now

File: python/ray/rllib/tuned_examples/pong-a3c.yaml [hunk 11]
```diff
@@ -2,7 +2,7 @@ pong-a3c:
     env: PongDeterministic-v4
     run: A3C
     config:
-        num_workers: 16
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/tuned_examples/pong-a3c.yaml]: Fixed

File: python/ray/rllib/utils/common_policy_evaluator.py [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+            self,
+            env_creator,
+            policy_graph,
+            tf_session_creator=None,
+            batch_steps=100,
+            batch_mode="truncate_episodes",
+            preprocessor_pref="rllib",
+            sample_async=False,
+            compress_observations=False,
+            consumer_buffer_size=0,
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/common_policy_evaluator.py]: Just removed for now.

File: python/ray/rllib/utils/common_policy_evaluator.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+                to be slightly off-policy.
+            compress_observations (bool): If true, compress the observations
+                returned.
+            consumer_buffer_size (int): If non-zero, buffers up to N sample
+                batches in-memory for use so that they can be retrieved
+                by multiple consumers. This only makes sense in multi-agent.
+            observation_filter (str): Name of observation filter to use.
+            registry (tune.Registry): Tune object registry. Pass in the value
+                from tune.registry.get_registry() if you're having trouble
+                resolving objects registered in tune.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/common_policy_evaluator.py]: Updated

File: python/ray/rllib/utils/common_policy_evaluator.py [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+            env_config (dict): Config to pass to the env creator.
+            model_config (dict): Config to use when creating the policy model.
+            policy_config (dict): Config to pass to the policy.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/common_policy_evaluator.py]: Done

File: python/ray/rllib/utils/tf_policy_graph.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+    def compute_actions(
+            self, obs_batch, state_batches=None, is_training=False):
+        state_batches = state_batches or []
+        assert len(self._state_inputs) == len(state_batches), \
+            (self._state_inputs, state_batches)
+        feed_dict = self.extra_compute_action_feed_dict()
+        feed_dict[self._obs_input] = obs_batch
+        feed_dict[self._is_training] = is_training
+        for ph, value in zip(self._state_inputs, state_batches):
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_graph.py]: Yeah, it's a list

File: python/ray/rllib/utils/tf_policy_graph.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    def extra_compute_grad_fetches(self):
+        return {}  # e.g, td error
+
+    def extra_apply_grad_feed_dict(self):
+        return {}
+
+    def extra_apply_grad_fetches(self):
+        return {}  # e.g., batch norm updates
+
+    def optimizer(self):
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_graph.py]: It's only called once.

**Comment by ericl** [comment]
@richardliaw this is ready for review

File: python/ray/rllib/a3c/a3c.py [hunk 17]
```diff
@@ -154,7 +147,8 @@ def _restore(self, checkpoint_path):
         ])
         self.local_evaluator.restore(extra_data["local_state"])
 
-    def compute_action(self, observation):
+    def compute_action(self, observation, state=[]):
```
Comment (robertnishihara) [inline python/ray/rllib/a3c/a3c.py]: better to avoid using mutable objects as default values, perhaps

`state=None` and `state = [] if state is None else state`

File: python/ray/rllib/dqn/dqn.py [hunk 18]
```diff
@@ -260,6 +237,7 @@ def _restore(self, checkpoint_path):
         self.num_target_updates = extra_data[3]
         self.last_target_update_ts = extra_data[4]
 
-    def compute_action(self, observation):
-        return self.local_evaluator.dqn_graph.act(
-            self.local_evaluator.sess, np.array(observation)[None], 0.0)[0]
+    def compute_action(self, observation, state=[]):
```
Comment (robertnishihara) [inline python/ray/rllib/dqn/dqn.py]: same comment here about default arguments

File: python/ray/rllib/optimizers/policy_optimizer.py [hunk 19]
```diff
@@ -38,18 +38,24 @@ def make(
...
             optimizer_config (dict): Keyword arguments to pass to the
                 optimizer class constructor.
         """
 
-        local_evaluator = evaluator_cls(*evaluator_args)
         remote_cls = ray.remote(**evaluator_resources)(evaluator_cls)
-        remote_evaluators = [
-            remote_cls.remote(*evaluator_args)
-            for _ in range(num_workers)]
+        if type(evaluator_args) is list:
```
Comment (robertnishihara) [inline python/ray/rllib/optimizers/policy_optimizer.py]: `isinstance(evaluator_args, list)`

File: python/ray/rllib/pg/pg.py [hunk 20]
```diff
@@ -50,34 +47,26 @@ def default_resource_request(cls, config):
...
-            episode_len_mean=avg_length,
-            timesteps_this_iter=timesteps,
-            info={})
-
-        return result
-
-    def compute_action(self, obs):
-        action, info = self.optimizer.local_evaluator.policy.compute(obs)
-        return action
+    def compute_action(self, observation, state=[]):
```
Comment (robertnishihara) [inline python/ray/rllib/pg/pg.py]: mutable default arg

File: python/ray/rllib/dqn/dqn.py (hunk 18 above)
Comment (ericl) [reply python/ray/rllib/dqn/dqn.py]: Fixed

File: python/ray/rllib/pg/pg.py (hunk 20 above)
Comment (ericl) [reply python/ray/rllib/pg/pg.py]: Fixed.

File: python/ray/rllib/a3c/a3c.py (hunk 17 above)
Comment (ericl) [reply python/ray/rllib/a3c/a3c.py]: Done

File: python/ray/rllib/utils/sampler.py [hunk 21]
```diff
@@ -6,76 +6,7 @@
 import threading
 from collections import namedtuple
 import numpy as np
```
Comment (richardliaw) [inline python/ray/rllib/utils/sampler.py]: space between ray and non-ray imports?

File: python/ray/rllib/utils/sampler.py [hunk 22]
```diff
@@ -252,37 +191,30 @@ def _env_runner(env, policy, num_local_steps, horizon, obs_filter):
...
 
             # Collect the experience.
-            rollout.add(
+            batch_builder.add_values(
                 obs=last_observation,
                 actions=action,
                 rewards=reward,
                 dones=terminal,
-                features=last_features,
                 new_obs=observation,
```
Comment (richardliaw) [inline python/ray/rllib/utils/sampler.py]: IDK where to leave this note, but we're actually doubling the number of states we need to send here (`observation` and `last_observation`). In a later optimization, we should consider addressing this (I can put this on TODOs)

File: python/ray/rllib/utils/policy_graph.py [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+        Arguments:
+            obs_batch (np.ndarray): batch of observations
+            state_batches (list): list of RNN state input batches, if any
+            is_training (bool): whether we are training the policy
+
+        Returns:
+            actions (np.ndarray): batch of output actions, with shape like
+                [BATCH_SIZE, ACTION_SHAPE].
+            state_outs (list): list of RNN state output batches, if any, with
+                shape like [STATE_SIZE, BATCH_SIZE].
```
Comment (richardliaw) [inline python/ray/rllib/utils/policy_graph.py]: why is BATCH after STATE here?

File: python/ray/rllib/utils/policy_graph.py [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+        Returns:
+            state (obj): Serialized local state.
+        """
+        return self.get_weights()
+
+    def set_state(self, state):
+        """Restores all local state.
+
+        Arguments:
+            state (obj): Serialized local state."""
```
Comment (richardliaw) [inline python/ray/rllib/utils/policy_graph.py]: nit: inconsistent quote placement

File: python/ray/rllib/ppo/loss.py [hunk 25]
```diff
@@ -49,7 +49,7 @@ def __init__(
         self.surr2 = tf.clip_by_value(self.ratio, 1 - config["clip_param"],
                                       1 + config["clip_param"]) * advantages
         self.surr = tf.minimum(self.surr1, self.surr2)
-        self.mean_policy_loss = tf.reduce_mean(-self.surr)
+        self.mean_policy_graph = tf.reduce_mean(-self.surr)
```
Comment (richardliaw) [inline python/ray/rllib/ppo/loss.py]: This naming change doesn't make sense.

**Review by richardliaw — CHANGES_REQUESTED** [review]
OK last comments - will merge by tonight after addressed.

File: python/ray/rllib/ppo/loss.py (hunk 25 above)
Comment (ericl) [reply python/ray/rllib/ppo/loss.py]: Fixed

File: python/ray/rllib/utils/tf_policy_graph.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_graph.py]: This is just not an issue. We are calling this ourselves, not the user, so it's impossible for them to screw it up.

**Comment by ericl** [comment]
Updated

**Review by richardliaw — APPROVED** [review]
